app/coffee/AuthHandler.coffee

```diff
@@ -0,0 +1,64 @@
+bcrypt = require("bcrypt")
+crypto = require("crypto")
+logger = require("logger-sharelatex")
+
+{ db } = require("./mongojs")
+Errors = require("./Errors")
+
+TOKEN_TTL_MS = 60 * 60 * 1000
+
+module.exports = AuthHandler =
+	login: (email, password, callback) ->
+		db.users.findOne {email: email.toLowerCase()}, (err, user) ->
+			if err
+				return callback(err)
+			if !user
+				return callback(new Errors.AuthenticationError("user not found"))
+			if !user.hashedPassword
+				return callback(new Errors.AuthenticationError("user does not have a password"))
+			bcrypt.compare password, user.hashedPassword, (err, match) ->
+				if err
+					return callback(err)
+				if !match
+					return callback(new Errors.AuthenticationError("incorrect password"))
+				callback(null, user._id.toString())
+
+	oneTimeLogin: (email, token, callback) ->
+		now = new Date()
+		normalizedEmail = email.toLowerCase()
+		query = {
+			email: normalizedEmail
+			token: token
+			expiresAt: { $gt: now }
+			usedAt: { $exists: false }
+		}
+		update = {
+			$set: { usedAt: now }
+		}
+		db.tokens.findAndModify { query: query, update: update }, (err, token) ->
```

Comment:
I'm good to make the collection name more explicit, like `oneTimeLoginTokens`, but it's not necessary to indicate which app the collection belongs to because the collections are in a separate database.

Also, read-only has `users` and `projects` collections that are different from the corresponding collections in web.

Reply:
Ah, I hadn't noticed it was in a separate `read-only` database at [settings.defaults.coffee#L33](https://github.com/overleaf/read-only/blob/36e962fdb8dc790de2c0ddab27c3014d93d27038/config/settings.defaults.coffee#L33) in the dev environment.